### ISSUE TITLE
UUID format unhyphenated in listings

### ISSUE THREAD

**Issue by sportluffi** [stratis-storage/stratis-cli#698]
PR #209 added UUID output fin the filesystem listing. However, the printed UUID is not valid as per [RFC 4122](https://tools.ietf.org/html/rfc4122#page-4) and should contain dashes. Fstab requires dashed UUIDs, so it is not possible to use the provided information from the `stratis fs list` command for persistent mounting - a roundtrip to blkid is required.

To fix this issue without potentially breaking UUID-depending logic, the simplest approach may be to add UUID formatting through the uuid library around [_logical.py#L170](https://github.com/stratis-storage/stratis-cli/blob/develop-2.3.0/src/stratis_cli/_actions/_logical.py#L170).

**Comment by mulkieran** [comment]
Please note that we officially discourage all users from programming against the CLI. stratisd provides a D-Bus interface with reliable and backwards compatible interfaces that we maintain; we do not promise to keep the output of the CLI stable.

Regarding the format of the UUID, stratisd uses UUIDs in its metadata where it uses the simple, rather than the hyphenated format. This is an obvious choice, as the hyphenated representation results in a value with a length that is not a power of 2, which is awkward for metadata. We have chosen to use the same format in all representations of UUIDs in the stratisd D-Bus API and in the stratisd logs. This choice is documented in the design doc, https://stratis-storage.github.io/StratisSoftwareDesign.pdf, but not in the D-Bus API doc, which is an oversight: https://stratis-storage.github.io/DBusAPIReference.pdf.

It's not really the case that in order to obtain a hyphenated UUID it is necessary to consult blkid. stratisd exports the correct value, if it is necessary to reformat that value with hyphens, and this can be the case in various contexts, then any language with any support for UUIDs will include support for reformatting the UUIDs. If sticking w/ shell scripting, a sed one-liner will do the trick as in the rough example below:

```
$ echo bfa83e18516c42fe887d6a64e7e31f80 | sed 's/\(^[a-f0-9]\{8\}\)\([a-f0-9]\{4\}\)\([a-f0-9\]\{4\}\)\([a-f0-9]\{4\}\)/\1-\2-\3-\4-/'
bfa83e18-516c-42fe-887d-6a64e7e31f80
```

**Comment by sportluffi** [comment]
Thank you for pointing me towards the design doc - and hinting, that the actual `blkid` roundtrip is not necessary. I can fully support the idea of using the simple UUID without dashes internally. I also do get, that the UUID is in fact the correct value - it's just formatted, or represented, in away that is usable for work within stratis - and not for e.g. Fstab.

However, it does feel like a less ideal way to involve other tools, even if its _just_ sed, to reformat the UUID to a usable UUID representation for working with  `/etc/fstab`. So while the internal UUID representation is a good design choice, I'd propose to add the dashed UUID represenation as an output. In my perspective, stratis should be able to deliver a useable represenation of the Filesystem identifier for further reference within the OS. I get, that the information can be obtained elsewhere in the wanted format - but the cli should be able to deliver the same information without hassle.

My goal is not to use the UUID in a scripted way, as I am looking into stratis as a beginner. It is merely used as a one-off operation when creating new filesystems, which doesn't occur often.

**Comment by mulkieran** [comment]
Ok! Essentially, when it comes to UUIDs, the CLI prints out exactly what is sent over the D-Bus. If we were to adjust the formatting of UUIDs we would most likely use some sort of global configuration option, as we display the UUID for a few commands.

**Comment by mulkieran** [comment]
We decided that we would make the hyphenated format the default, but use a global configuration option to allow users to choose the unhyphenated format.

**Comment by mulkieran** [comment]
There are a couple of parts to this:
0. Figure out how to change the format of a single UUID. Use the uuid Python library, there will certainly be something there.
1. Add the global ```--unhyphenated-UUIDs``` flag. Maybe we can come up w/ something shorter that still makes sense. The one other global flag we have is ```--propagate```, as an example.
2. Figure out how to properly observe this flag whenever printing out UUIDs. This is definitely the hard part. The constraint we have is that ```print_table``` expects its fields to be in their final form, so it can align those columns properly. There are many possibilities here. The challenge is to come up with something that is reasonably maintainable.

**Comment by lleshchi** [comment]
@mulkieran Do you want the flag to be an argument of the `fs` command or the `list` command?
Option 1: `stratis fs --unhyphenated-uuids` and `stratis fs --unhyphenated-uuids list`
Option 2: `stratis fs list --unhyphenated-uuids` which I believe will prevent use of the unhyphenated format with the shortcut `fs` command for listing

Perhaps there is a better way to do this, but I believe providing this flag as an argument for both `fs` and `list` would not be possible unless they are stored under different attributes.

**Comment by mulkieran** [comment]
> @mulkieran Do you want the flag to be an argument of the `fs` command or the `list` command?
> Option 1: `stratis fs --unhyphenated-uuids` and `stratis fs --unhyphenated-uuids list`
> Option 2: `stratis fs list --unhyphenated-uuids` which I believe will prevent use of the unhyphenated format with the shortcut `fs` command for listing
> 
> Perhaps there is a better way to do this, but I believe providing this flag as an argument for both `fs` and `list` would not be possible unless they are stored under different attributes.

@lleshchi What I was thinking was to set the option up as a global, just like '--propagate'. So it would be invoked
```> stratis --unhyphenated-uuids pool list``` 
or
```> stratis --unhyphenated-uuids fs list```
for example. That way it would cover our expectations for UUIDs in a more global way.

Another option is to attach it to the sub-commands, as in your example, so it would be invoked:

```> stratis fs --unhyphenated-uuids list```
or
```> stratis pool --unhyphenated-uuids list```

And the third is to attach it to the specific command, also as in your example above. You may well be correct that the third option, as, 
```> stratis pool list --unhyphenated-uuids``` is the most difficult to implement, due to the defaulting behavior for all the subcommands.

I'm still liking the first, global option, best, for various reasons. stratis runs only one command at a time, so the global option will always do what the user wants (unless they want some crazy complicated thing with differently formatted UUIDs in the output of a single command, which would be very odd and wouldn't be satisfied by any of the ideas we've tried so far). it is also probably the simplest to implement.

@drckeefe What do you think?

**Comment by mulkieran** [comment]
@drckeefe is agreeable to the global option.

**Comment by mulkieran** [comment]
Closed by #705.
